### PR TITLE
Add `DataFrame` and `Series` `median` method

### PR DESCRIPTION
**Pull request by jrbourbeau**
This PR adds new `median` and `median_approximate` methods to `dask.dataframe`.

Previously there was hesitation around using approximate algorithms in a `DataFrame.median` method, which could lead to confusion for users who are expecting `dask` and `pandas` to always return the same result. This PR proposes we use `median` in cases where we can provide an exact median calculation (i.e. when `axis=1` or when there's only a single partition in the `DataFrame`/`Series`). Otherwise, we raise an informative error directing users to `median_approximate` and (internally) use the existing approximate `quantiles` implementation. 

Closes https://github.com/dask/dask/issues/4362, supersedes https://github.com/dask/dask/pull/3819

TODO:

- [x] Add median methods to `dd.Series`
- [ ] ~~Add median to groupby aggregations~~ To reduce the scope of this PR, let's handle this in as a follow-up 
- [x] Update Dask DataFrame API docs

### CODE REVIEW

File: dask/dataframe/tests/test_dataframe.py [hunk 1]
```diff
@@ -4307,6 +4334,51 @@ def test_dataframe_mode():
     assert_eq(ddf.mode(), df.mode(), check_index=False)
 
 
+def test_median():
+    df = pd.DataFrame({"x": [1, 2, 3, 4, 5], "y": [1.1, 2.2, 3.3, 4.4, 5.5]})
+    ddf = dd.from_pandas(df, npartitions=3)
+
+    # Exact medians work when `axis=1` or when there's only have a single partition
+    # TODO: shouldn't need to specify `check_names=False` below, but names currently don't match
+    assert_eq(ddf.median(axis=1), df.median(axis=1), check_names=False)
```
Comment (ian-r-rose) [inline dask/dataframe/tests/test_dataframe.py:4343]: I believe that we can remove the `check_names=False` by chaining a `.rename(None)` onto the dataframe median functions (the series ones are just scalars, so don't have names)

Comment (jrbourbeau) [reply dask/dataframe/tests/test_dataframe.py]: Thanks for the suggestion @ian-r-rose. I went ahead and added `.rename(None)` and removed the `check_names=False` lines (and corresponding `TODO` comments). We don't often need to use `rename` so it still feels like there's something deeper happening with the name. That said, I don't want to block progress on this point -- adding a `rename` call is a valid thing to do

Comment (ian-r-rose) [reply dask/dataframe/tests/test_dataframe.py]: I don't think there's really anything weird going on with the series name: it's just a difference between the behaviors of `quantile` and `median`. With pandas, each quantile series gets the name of the quantile, but median does not:

```python
In [1]: import pandas

In [2]: df = pandas.DataFrame({"a": [1,2,3,4], "b": [5,6,7,8]})

In [3]: df.median(axis=1)
Out[3]: 
0    3.0
1    4.0
2    5.0
3    6.0
dtype: float64

In [4]: df.quantile(axis=1)
Out[4]: 
0    3.0
1    4.0
2    5.0
3    6.0
Name: 0.5, dtype: float64
```

Comment (jrbourbeau) [reply dask/dataframe/tests/test_dataframe.py]: Ah, great

File: dask/dataframe/tests/test_dataframe.py [hunk 2]
```diff
@@ -4307,6 +4334,53 @@ def test_dataframe_mode():
...
 
 
+def test_median():
+    df = pd.DataFrame({"x": [1, 2, 3, 4, 5], "y": [1.1, 2.2, 3.3, 4.4, 5.5]})
+    ddf = dd.from_pandas(df, npartitions=3)
+
+    # Exact medians work when `axis=1` or when there's only have a single partition
+    assert_eq(ddf.median(axis=1), df.median(axis=1))
+    ddf_single = dd.from_pandas(df, npartitions=1)
+    assert_eq(ddf_single.median(axis=1), df.median(axis=1))
```
Comment (ian-r-rose) [inline dask/dataframe/tests/test_dataframe.py:4344]: Nit: we probably want to text `axis=0` for `ddf_single.median`

**Review by ian-r-rose — COMMENTED** [review]
One small suggestion, otherwise this looks good to me

File: dask/dataframe/tests/test_dataframe.py (hunk 2 above)
Comment (jrbourbeau) [reply dask/dataframe/tests/test_dataframe.py:4344]: Good catch! Just pushed a commit to test that case

**Review by ian-r-rose — APPROVED** [review]

**Comment by ian-r-rose** [comment]
Test failure is a known flaky one #8816

**Comment by jrbourbeau** [comment]
Thanks for reviewing @ian-r-rose!

**Pull request merged**